lib-model/lib-model-consume/src/main/java/com/uber/nullaway/libmodel/LibModelInfoExtractor.java

```diff
@@ -0,0 +1,303 @@
+        .getSourceRoots()
+        .forEach(
+            sourceRoot -> {
+              try {
+                sourceRoot.parse("", mc);
+              } catch (IOException e) {
+                System.err.println("IOException: " + e);
+              }
+            });
+  }
+
+  public static void writeToAstubx(String outputPath) {
+    Map<String, String> importedAnnotations =
+        ImmutableMap.<String, String>builder()
+            .put("Nonnull", "javax.annotation.Nonnull")
+            .put("Nullable", "javax.annotation.Nullable")
+            .build();
+    DataOutputStream dos;
+    try {
+      Path opPath = Paths.get(outputPath);
+      Files.createDirectories(opPath.getParent());
+      dos = new DataOutputStream(Files.newOutputStream(opPath));
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    if (!methodRecords.isEmpty()) {
+      try {
+        StubxFileWriter.write(
+            dos, importedAnnotations, new HashMap<>(), new HashMap<>(), methodRecords);
+        dos.close();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  public static Path dirnameToPath(String dir) {
+    File f = new File(dir);
+    if (!f.exists()) {
+      System.err.printf("Directory %s (%s) does not exist.%n", dir, f);
+      System.exit(1);
+    }
+    if (!f.isDirectory()) {
+      System.err.printf("Not a directory: %s (%s).%n", dir, f);
+      System.exit(1);
+    }
+    String absoluteDir = f.getAbsolutePath();
+    if (absoluteDir.endsWith("/.")) {
+      absoluteDir = absoluteDir.substring(0, absoluteDir.length() - 2);
+    }
+    return Paths.get(absoluteDir);
+  }
+
+  /** Callback to process each Java file; see class documentation for details. */
+  private static class MinimizerCallback implements SourceRoot.Callback {
+
+    /** The visitor instance. */
+    private final CompilationUnitVisitor mv;
+
+    /** Create a MinimizerCallback instance. */
+    public MinimizerCallback() {
+      this.mv = new CompilationUnitVisitor();
+    }
+
+    @Override
+    public Result process(Path localPath, Path absolutePath, ParseResult<CompilationUnit> result) {
+      Result res = Result.SAVE;
+      Optional<CompilationUnit> opt = result.getResult();
+      if (opt.isPresent()) {
+        CompilationUnit cu = opt.get();
+        visitAll(cu);
+      }
+      return res;
+    }
+
+    public void visitAll(Node rootNode) {
+      if (rootNode == null) {
+        return;
+      }
+      ArrayDeque<Node> stack = new ArrayDeque<>();
+      stack.addFirst(rootNode);
+      while (!stack.isEmpty()) {
+        Node current = stack.removeFirst();
+        current.accept(mv, null);
+        List<Node> children = current.getChildNodes();
+        if (children != null) {
+          for (Node child : children) {
+            stack.addFirst(child);
+          }
+        }
+      }
+    }
+  }
+
+  private static class CompilationUnitVisitor extends ModifierVisitor<Void> {
```

Comment:
This should be named something like `AnnotationCollectionVisitor`.  And I think it should just extend `VoidVisitorAdapter<Void>` since we don't modify anything and don't rely on the return values.

Reply:
Yes I agree, I have changed it.